pkg/kubelet/logs/container_log_manager.go

```diff
@@ -0,0 +1,401 @@
+				continue
+			}
+			// The container log should be recovered.
+			info, err = os.Stat(path)
+			if err != nil {
+				glog.Errorf("Failed to stat container log %q after reopen: %v", path, err)
+				continue
+			}
+		}
+		if info.Size() < c.policy.MaxSize {
+			continue
+		}
+		// Perform log rotation.
+		if err := c.rotateLog(id, path); err != nil {
+			glog.Errorf("Failed to rotate log %q for container %q: %v", path, id, err)
+			continue
+		}
+	}
+	return nil
+}
+
+func (c *containerLogManager) rotateLog(id, log string) error {
+	// pattern is used to match all rotated files.
+	pattern := fmt.Sprintf("%s.*", log)
+	logs, err := filepath.Glob(pattern)
+	if err != nil {
+		return fmt.Errorf("failed to list all log files with pattern %q: %v", pattern, err)
+	}
+
+	logs, err = c.cleanupUnusedLogs(logs)
+	if err != nil {
+		return fmt.Errorf("failed to cleanup logs: %v", err)
+	}
+
+	logs, err = c.removeExcessLogs(logs)
+	if err != nil {
+		return fmt.Errorf("failed to remove excess logs: %v", err)
+	}
+
+	// Compress uncompressed log files.
+	for _, l := range logs {
+		if strings.HasSuffix(l, compressSuffix) {
+			continue
+		}
+		if err := c.compressLog(l); err != nil {
+			return fmt.Errorf("failed to compress log %q: %v", l, err)
+		}
+	}
+
+	if err := c.rotateLatestLog(id, log); err != nil {
+		return fmt.Errorf("failed to rotate log %q: %v", log, err)
+	}
+
+	return nil
+}
+
+// cleanupUnusedLogs cleans up temporary or unused log files generated by previous log rotation
+// failure.
+// logs passed in must be in increasing order.
+func (c *containerLogManager) cleanupUnusedLogs(logs []string) ([]string, error) {
+	inuse, unused := filterUnusedLogs(logs)
+	for _, l := range unused {
+		if err := os.Remove(l); err != nil {
+			return nil, fmt.Errorf("failed to remove unused log %q: %v", l, err)
+		}
+	}
+	return inuse, nil
+}
+
+// filterUnusedLogs splits logs into 2 groups, the 1st group is in used logs,
+// the second group is unused logs.
+func filterUnusedLogs(logs []string) ([]string, []string) {
+	var inuse []string
+	var unused []string
+
+	for _, l := range logs {
+		// All temporary files are not in use.
+		if strings.HasSuffix(l, tmpSuffix) {
+			unused = append(unused, l)
+			continue
+		}
+		// All compresed logs are in use.
+		if strings.HasSuffix(l, compressSuffix) {
+			inuse = append(inuse, l)
+			continue
+		}
+		// Files has already been compressed are not in use.
+		var found bool
+		for _, another := range logs {
+			if l+compressSuffix == another {
+				found = true
+				break
+			}
+		}
+		if found {
+			unused = append(unused, l)
+			continue
+		}
+		inuse = append(inuse, l)
+	}
+	return inuse, unused
+}
+
+// removeExcessLogs removes old logs to make sure there are only at most MaxFiles log files.
+// logs passed in must be in increasing order.
+func (c *containerLogManager) removeExcessLogs(logs []string) ([]string, error) {
+	// Sort log files in oldest to newest order.
+	sort.Strings(logs)
```

Comment:
I thought logs were already required to be in sorted order when being passed to `filterUnusedLogs`?

Reply:
They should still be in that order.